src/qp/metrics/base_metric_classes.py

```diff
@@ -0,0 +1,167 @@
+import enum
+
+from abc import ABC
+
+class MetricInputType(enum.Enum):
+    """Defines the various combinations of input types that metric classes accept.
+    """
+
+    unknown = -1
+
+    # A single qp.Ensemble
+    single_ensemble = 0
+
+    # A distribution, or collection of distributions for estimate(s) and a
+    # single value, or collection of values for reference(s)
+    dist_to_point = 1
+
+    # A distribution, or collection of distributions for estimate(s) and a
+    # distribution, or collection of distributions for references(s)
+    dist_to_dist = 2
+
+    # A single value, or collection of values for estimate(s) and a
+    # single value, or collection of values for reference(s).
+    point_to_point = 3
+
+    # A single value, or collection of values for estimate(s) and a
+    # distribution, or collection of distributions for reference(s).
+    point_to_dist = 4
+
+
+    def uses_distribution_for_estimate(self) -> bool:
+        return self in [
+            MetricInputType.single_ensemble,
+            MetricInputType.dist_to_point,
+            MetricInputType.dist_to_dist,
+        ]
+
+    def uses_distribution_for_reference(self) -> bool:
+        return self in [
+            MetricInputType.dist_to_dist,
+            MetricInputType.point_to_dist,
+        ]
+
+    def uses_point_for_estimate(self) -> bool:
+        return self in [
+            MetricInputType.point_to_dist,
+            MetricInputType.point_to_point,
+        ]
+
+    def uses_point_for_reference(self) -> bool:
+        return self in [
+            MetricInputType.dist_to_point,
+            MetricInputType.point_to_point,
+        ]
+
+
+class MetricOuputType(enum.Enum):
+    """Defines the various output types that metric classes can return.
+    """
+
+    unknown = -1
+
+    # The metric produces a single value for all input
+    single_value = 0
+
+    # The metric produces a single distribution for all input
+    single_distribution = 1
+
+    # The metric produces a value for each input distribution
+    one_value_per_distribution = 2
+
+
+class BaseMetric(ABC):
```

Comment:
Maybe some docstrings here and in the intermediate level base classes.

Reply:
Done